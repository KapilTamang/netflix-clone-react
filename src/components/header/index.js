import React, { useState } from 'react';
import { Link as ReactRouterLink } from 'react-router-dom';
import {
	Background,
	Container,
	Group,
	Logo,
	ButtonLink,
	Feature,
	FeatureCallOut,
	Text,
	Link,
	Image,
	Profile,
	Dropdown,
	Search,
	SearchIcon,
	SearchInput,
	PlayButton,
} from './styles/header';

export default function Header({ bg = true, children, ...restProps }) {
	return bg ? (
		<Background {...restProps} data-testid="header-bg">
			{children}
		</Background>
	) : (
		children
	);
}

Header.Feature = function HeaderFeature({ children, ...restProps }) {
	return <Feature {...restProps}>{children}</Feature>;
};

Header.FeatureCallOut = function HeaderFeatureCallOut({
	children,
	...restProps
}) {
	return <FeatureCallOut {...restProps}>{children}</FeatureCallOut>;
};

Header.PlayButton = function HeaderPlayButton({ children, ...restProps }) {
	return <PlayButton {...restProps}>{children}</PlayButton>;
};
Header.Text = function HeaderText({ children, ...restProps }) {
	return <Text {...restProps}>{children}</Text>;
};

Header.TextLink = function HeaderTextLink({ children, ...restProps }) {
	return <Link {...restProps}>{children}</Link>;
};

Header.Frame = function HeaderFrame({ children, ...restProps }) {
	return <Container {...restProps}>{children}</Container>;
};

Header.Group = function HeaderGroup({ children, ...restProps }) {
	return <Group {...restProps}>{children}</Group>;
};

Header.Logo = function HeaderLogo({ to, ...restProps }) {
	return (
		<ReactRouterLink to={to}>
			<Logo {...restProps} />
		</ReactRouterLink>
	);
};

Header.ButtonLink = function HeaderButtonLink({ children, ...restProps }) {
	return <ButtonLink {...restProps}>{children}</ButtonLink>;
};

Header.Image = function HeaderImage({ src, children, ...restProps }) {
	return (
		<Image {...restProps} src={`/images/users/${src}.png`}>
			{children}
		</Image>
	);
};

Header.Profile = function HeaderProfile({ children, ...restProps }) {
	return <Profile {...restProps}>{children}</Profile>;
};

Header.Dropdown = function HeaderDropdown({ children, ...restProps }) {
	return <Dropdown {...restProps}>{children}</Dropdown>;
};

Header.Search = function HeaderSearch({
	searchTerm,
	setSearchTerm,
	...restProps
}) {
	const [searchActive, setSearchActive] = useState(false);

	return (
		<Search {...restProps}>
			<SearchIcon
				onClick={() => setSearchActive(!searchActive)}
				data-testid="search-click"
			>
				<img
					src="/images/icons/search.png"
					alt="Search"
					data-testid="search-icon"
				/>
			</SearchIcon>
			<SearchInput
				value={searchTerm}
				onChange={(e) => setSearchTerm(e.target.value)}
				placeholder="Search films and series"
				active={searchActive}
				data-testid="search-input"
			></SearchInput>
		</Search>
	);
};
