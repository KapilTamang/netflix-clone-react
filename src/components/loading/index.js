import React from 'react';
import { Spinner, LockBody, ReleaseBody, Image } from './styles/loading';

export default function Loading({ src, restProps }) {
	return (
		<Spinner {...restProps}>
			<LockBody />
			<Image src={`/images/users/${src}.png`} data-testid="loading-image" />
		</Spinner>
	);
}

Loading.ReleaseBody = function LoadingReleaseBody() {
	return <ReleaseBody />;
};
