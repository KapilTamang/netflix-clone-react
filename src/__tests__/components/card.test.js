import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Card, Player } from '../../components';

const category = 'films';

const slideRows = [
	{
		title: 'Thriller',
		data: [
			{
				description: 'Joker description',
				genre: 'thriller',
				docId: '14YSyR6OH25BSQWQ5Xzx',
				id: '6a4bf6a5-b9f6-4eeb-bbd3-80cba3b9b645',
				maturity: '15',
				slug: 'joker',
				title: 'Joker',
			},
		],
	},
	{
		title: 'Children',
		data: [
			{
				description: 'Frozen description',
				genre: 'children',
				docId: 'MebFh7Ve4IFBMDl9PHSt',
				id: '845507bf-f3cc-40de-b633-485a06449a27',
				maturity: '0',
				slug: 'frozen',
				title: 'Frozen',
			},
		],
	},
];

describe('<Card/>', () => {
	it('renders <Card/> with populated data.', () => {
		const { container, getByText } = render(
			<Card.Group>
				{slideRows.map((slideItem) => (
					<Card key={`${category}-${slideItem.title.toLowerCase()}`}>
						<Card.Title>{slideItem.title}</Card.Title>
						<Card.Entities>
							{slideItem.data.map((item) => (
								<Card.Item key={item.docId} item={item}>
									<Card.Image
										src={`/images/${category}/${item.genre}/${item.slug}/small.jpg`}
									/>
									<Card.Meta>
										<Card.SubTitle>{item.title}</Card.SubTitle>
										<Card.Text>{item.description}</Card.Text>
									</Card.Meta>
								</Card.Item>
							))}
						</Card.Entities>
						<Card.Feature category={category}>
							<Player>
								<Player.Button />
								<Player.Video src="/videos/test.mp4" />
							</Player>
						</Card.Feature>
					</Card>
				))}
			</Card.Group>
		);

		expect(getByText('Thriller')).toBeTruthy();
		expect(getByText('Joker')).toBeTruthy();
		expect(getByText('Joker description')).toBeTruthy();

		expect(getByText('Children')).toBeTruthy();
		expect(getByText('Frozen')).toBeTruthy();
		expect(getByText('Frozen description')).toBeTruthy();

		expect(container.firstChild).toMatchSnapshot();
	});

	it('renders the <Card/> and toggles the card feature', () => {
		const { container, queryByText, queryByTestId, getByAltText } = render(
			<Card.Group>
				{slideRows.map((slideItem) => (
					<Card key={`${category}-${slideItem.title.toLowerCase()}`}>
						<Card.Title>{slideItem.title}</Card.Title>
						<Card.Entities>
							{slideItem.data.map((item) => (
								<Card.Item
									key={item.docId}
									item={item}
									data-testid={`${item.slug}-item-feature`}
								>
									<Card.Image
										src={`/images/${category}/${item.genre}/${item.slug}/small.jpg`}
									/>
									<Card.Meta>
										<Card.SubTitle>{item.title}</Card.SubTitle>
										<Card.Text>{item.description}</Card.Text>
									</Card.Meta>
								</Card.Item>
							))}
						</Card.Entities>
						<Card.Feature category={category}>
							<Player>
								<Player.Button />
								<Player.Video src="/videos/test.mp4" />
							</Player>
						</Card.Feature>
					</Card>
				))}
			</Card.Group>
		);

		expect(queryByTestId('player')).toBeFalsy();
		fireEvent.click(queryByTestId('joker-item-feature'));
		expect(queryByText('15')).toBeTruthy();

		fireEvent.click(getByAltText('Close'));
		expect(queryByText('15')).toBeFalsy();

		fireEvent.click(queryByTestId('frozen-item-feature'));
		expect(queryByText('PG')).toBeTruthy();

		fireEvent.click(getByAltText('Close'));
		expect(queryByText('PG')).toBeFalsy();
	});
});
