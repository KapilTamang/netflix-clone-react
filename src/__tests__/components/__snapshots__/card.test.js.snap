// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card/> renders <Card/> with populated data. 1`] = `
<div
  class="card__Group-sc-17gu7hr-2 hhuDkt"
>
  <div
    class="card__Container-sc-17gu7hr-1 hoOxEJ"
  >
    <p
      class="card__Title-sc-17gu7hr-0 kQqYmG"
    >
      Thriller
    </p>
    <div
      class="card__Entities-sc-17gu7hr-12 bTuGXH"
    >
      <div
        class="card__Item-sc-17gu7hr-13 dfWZXi"
      >
        <img
          class="card__Image-sc-17gu7hr-14 evhgOr"
          src="/images/films/thriller/joker/small.jpg"
        />
        <div
          class="card__Meta-sc-17gu7hr-11 kINZZP"
        >
          <p
            class="card__SubTitle-sc-17gu7hr-3 cfxsNq"
          >
            Joker
          </p>
          <p
            class="card__Text-sc-17gu7hr-4 eRvMKP"
          >
            Joker description
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card__Container-sc-17gu7hr-1 hoOxEJ"
  >
    <p
      class="card__Title-sc-17gu7hr-0 kQqYmG"
    >
      Children
    </p>
    <div
      class="card__Entities-sc-17gu7hr-12 bTuGXH"
    >
      <div
        class="card__Item-sc-17gu7hr-13 dfWZXi"
      >
        <img
          class="card__Image-sc-17gu7hr-14 evhgOr"
          src="/images/films/children/frozen/small.jpg"
        />
        <div
          class="card__Meta-sc-17gu7hr-11 kINZZP"
        >
          <p
            class="card__SubTitle-sc-17gu7hr-3 cfxsNq"
          >
            Frozen
          </p>
          <p
            class="card__Text-sc-17gu7hr-4 eRvMKP"
          >
            Frozen description
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
